From 6d52b4dfa478028ac7fbf6df305093d60ecd39a5 Mon Sep 17 00:00:00 2001
From: Jan Sowinski <jan.sowinski@intel.com>
Date: Thu, 5 Dec 2024 12:34:53 +0100
Subject: [PATCH] i2c: ast: send stop on transaction failure

Added explicit generation of stop event when I2C transaction fails or if
address was not ACKed (e.g.: while scanning bus with probe command).
Prior to this change SCL remained low after failed transaction until
next transaction.

Similar behavior of generating stop event on failures is done
in kernel driver.

Signed-off-by: Jan Sowinski <jan.sowinski@intel.com>
Upstream-Status: Pending
---
 drivers/i2c/ast_i2c.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/drivers/i2c/ast_i2c.c b/drivers/i2c/ast_i2c.c
index 04759cc1d4..0858a14395 100644
--- a/drivers/i2c/ast_i2c.c
+++ b/drivers/i2c/ast_i2c.c
@@ -198,7 +198,7 @@ static int ast_i2c_read_data(struct udevice *dev, u8 chip_addr, u8 *buffer,
 
 	ret = ast_i2c_start_txn(dev, (chip_addr << 1) | I2C_M_RD);
 	if (ret < 0)
-		return ret;
+		goto err_send_stop;
 
 	for (; len > 0; len--, buffer++) {
 		if (len == 1)
@@ -206,7 +206,7 @@ static int ast_i2c_read_data(struct udevice *dev, u8 chip_addr, u8 *buffer,
 		writel(i2c_cmd, &priv->regs->csr);
 		ret = ast_i2c_wait_isr(dev, I2CD_INTR_RX_DONE);
 		if (ret < 0)
-			return ret;
+			goto err_send_stop;
 		*buffer = (readl(&priv->regs->trbbr) & I2CD_RX_DATA_MASK)
 				>> I2CD_RX_DATA_SHIFT;
 	}
@@ -216,6 +216,10 @@ static int ast_i2c_read_data(struct udevice *dev, u8 chip_addr, u8 *buffer,
 		return ast_i2c_send_stop(dev);
 
 	return 0;
+
+err_send_stop:
+	ast_i2c_send_stop(dev);
+	return ret;
 }
 
 static int ast_i2c_write_data(struct udevice *dev, u8 chip_addr, u8
@@ -226,20 +230,24 @@ static int ast_i2c_write_data(struct udevice *dev, u8 chip_addr, u8
 
 	ret = ast_i2c_start_txn(dev, (chip_addr << 1));
 	if (ret < 0)
-		return ret;
+		goto err_send_stop;
 
 	for (; len > 0; len--, buffer++) {
 		writel(*buffer, &priv->regs->trbbr);
 		writel(I2CD_M_TX_CMD, &priv->regs->csr);
 		ret = ast_i2c_wait_tx(dev);
 		if (ret < 0)
-			return ret;
+			goto err_send_stop;
 	}
 
 	if (send_stop)
 		return ast_i2c_send_stop(dev);
 
 	return 0;
+
+err_send_stop:
+	ast_i2c_send_stop(dev);
+	return ret;
 }
 
 static int ast_i2c_deblock(struct udevice *dev)
-- 
2.34.1

